pageserver/examples/sketch_decoupled_pageservice.rs

```diff
@@ -0,0 +1,294 @@
+//! # Memory Management
+//!
+//! Before Timeline shutdown we have a reference cycle
+//! ```
+//! Timeline::handlers =strong=> HandlerTimeline =strong=> Timeline
+//! ```
+//!
+//! We do this so the hot path need only do one, uncontended, atomic increment, i.e.,
+//! the Weak<HandlerTimeline>.
+//!
+//! Timeline::shutdown breaks the reference cycle.
+//!
+//! Conn shutdown removes the Arc<HandlerTimeline> from Timeline::handlers.
+
+use tracing::{info, Instrument};
+
+struct Timeline {
+    tenant_id: TenantId,
+    shard_id: ShardId,
+    timeline_id: TimelineId,
+
+    gate: utils::sync::gate::Gate,
+    // None = shutting down
+    handlers: Mutex<Option<Vec<Arc<HandlerTimeline>>>>,
+}
+
+impl Timeline {
+    async fn shutdown(&self) {
+        info!("Shutting down Timeline");
+        scopeguard::defer!(info!("Timeline shut down"));
+        // prevent new handlers from getting created
+        let handlers = self.handlers.lock().expect("mutex poisoned").take();
+        if let Some(handlers) = handlers.as_ref() {
+            info!("Dropping {} Arc<HandlerTimeline>s", handlers.len());
+        }
+        drop(handlers);
+        self.gate.close().await;
+    }
+    fn getpage(&self, key: usize) {
+        info!("get {key}");
+    }
+    fn rel_size(&self) {
+        if self.shard_id != 0 {
+            info!("rel_size not supported on shard {}", self.shard_id);
+            return;
+        }
+        info!("rel_size");
+    }
+}
+
+impl Drop for Timeline {
+    fn drop(&mut self) {
+        info!("Dropping Timeline");
+    }
+}
+
+type TenantId = u64;
+type ShardId = usize;
+type TimelineId = String;
+
+struct PageServerHandler {
+    cache: HashMap<(TenantId, ShardId, TimelineId), Weak<HandlerTimeline>>,
+}
+
+impl Drop for PageServerHandler {
+    fn drop(&mut self) {
+        info!("Dropping PageServerHandler, unregistering its handlers from cached Timeline");
+        for (_, cached) in self.cache.drain() {
+            if let Some(cached) = cached.upgrade() {
+                let mut lock_guard = cached.timeline.handlers.lock().expect("mutex poisoned");
+                if let Some(handlers) = &mut *lock_guard {
+                    let pre = handlers.len();
+                    handlers.retain(|handler| !Arc::ptr_eq(handler, &cached));
+                    let post = handlers.len();
+                    info!("Removed {} handlers", pre - post);
```

Comment:
this would be 0 or 1 and be done N times, better to collect all upgradeable handlers and retain once.